src/Elmish/Dispatch.purs

```diff
@@ -1,45 +1,90 @@
--- |
--- |     textarea
--- |       { value: state.text
--- |       , onChange: dispatch <?| \e -> TextChanged <$> eventTargetValue e
--- |       , onMouseDown: dispatch <| \_ -> TextareaClicked
--- |       }
--- |
--- |       where
--- |         eventTargetValue = readForeign >=> lookup "target" >=> readForeign >=> lookup "value"
--- |
-handle :: forall arg msg. Dispatch msg -> (arg -> msg) -> E.EffectFn1 arg Unit
-handle dispatch fn = E.mkEffectFn1 $ dispatch <<< fn
-
--- | Same as `handle`, but dispatches a message optionally. See comments on
--- | `handle` for an example.
-handleMaybe :: forall arg msg. Dispatch msg -> (arg -> Maybe msg) -> E.EffectFn1 arg Unit
-handleMaybe dispatch fn = E.mkEffectFn1 $ maybe (pure unit) dispatch <<< fn
+class Handle msg event f where
+    -- | A convenience function to make construction of event handlers with
+    -- | arguments (i.e. `EffectFn1`) a bit shorter. The first parameter is a
+    -- | `Dispatch`. The second parameter can be either a message or a function
+    -- | from the event object to a message.
+    -- |
+    -- | Expected usage for this function is in its operator form `<|`
+    -- |
+    -- |     textarea
+    -- |       { value: state.text
+    -- |       , onChange: dispatch <| \e -> TextAreaChanged (E.textareaText e)
+    -- |       , onMouseDown: dispatch <| TextareaClicked
+    -- |       }
+    -- |
+    handle :: Dispatch msg -> f -> E.EffectFn1 event Unit
+
+class HandleMaybe msg event f where
+    -- | A variant of `handle` (aka `<|`) that allows to dispatch a message or
+    -- | not conditionally via returning a `Maybe message`.
+    -- |
+    -- | Expected usage for this function is in its operator form `<?|`
+    -- |
+    -- |     div
+    -- |       { onMouseDown: dispatch <?| \(E.MouseEvent e) ->
+    -- |           if e.ctrlKey then Just ClickedWithControl else Nothing
+    -- |       }
+    -- |
+    handleMaybe :: Dispatch msg -> f -> E.EffectFn1 event Unit
+
+instance (TypeEquals msg output, TypeEquals event input) => Handle msg event (input -> output) where
+    handle dispatch f = E.mkEffectFn1 $ dispatch <<< coerce <<< f <<< coerce
+else instance TypeEquals msg output => Handle msg event output where
+    handle dispatch msg = E.mkEffectFn1 \_ -> dispatch $ coerce msg
+
+instance (TypeEquals msg output, TypeEquals event input) => HandleMaybe msg event (input -> Maybe output) where
+    handleMaybe dispatch f = E.mkEffectFn1 $ maybe (pure unit) dispatch <<< coerce <<< f <<< coerce
+else instance TypeEquals msg output => HandleMaybe msg event (Maybe output) where
+    handleMaybe dispatch msg = E.mkEffectFn1 \_ -> maybe (pure unit) dispatch $ coerce msg
+
+class HandleEffect event f where
+    -- | An escape-hatch way to create an event handler for when neither
+    -- | `handle` nor `handleMaybe` are appropriate, which usually happens when
+    -- | the event handler must do something else, besides dispatching a
+    -- | message.
```

Comment:
Or dispatch multiple messages

Reply:
Well, I really don't want to encourage that pattern, so I purposely left it out.